src/protocol.rs

```diff
@@ -32,65 +33,62 @@ const DISCRIMINANT_LEN: usize = 1;
 
 type Result<T, E = Error> = std::result::Result<T, E>;
 
-pub async fn spawn(port: u16) -> std::io::Result<Vec<tokio::task::JoinHandle<()>>> {
-    let ipv4: SocketAddr = (std::net::Ipv4Addr::UNSPECIFIED, port).into();
-    let ipv6: SocketAddr = (std::net::Ipv6Addr::UNSPECIFIED, port).into();
-    let mut tasks = Vec::new();
-
-    tracing::info!(%port, "spawning IPv4 and IPv6 QCMP sockets");
-    for (address, socket) in [
-        (ipv4, tokio::net::UdpSocket::bind(ipv4).await?),
-        (ipv6, tokio::net::UdpSocket::bind(ipv6).await?),
-    ] {
-        tasks.push(tokio::spawn(
-            async move {
-                // Initialize a buffer for the UDP packet. We use the maximum size of a UDP
-                // packet, which is the maximum value of 16 a bit integer.
-                let mut input_buf = vec![0; 1 << 16];
-                let mut output_buf = Vec::new();
-                loop {
-                    tracing::debug!(%address, "awaiting qcmp packets");
-
-                    match socket.recv_from(&mut input_buf).await {
-                        Ok((size, source)) => {
-                            let received_at = chrono::Utc::now().timestamp_nanos();
-                            let contents = &input_buf[..size];
-
-                            let command = match Protocol::parse(contents) {
-                                Ok(Some(command)) => command,
-                                Ok(None) => {
-                                    tracing::debug!("rejected non-qcmp packet");
-                                    continue;
-                                }
-                                Err(error) => {
-                                    tracing::debug!(%error, "rejected malformed packet");
-                                    continue;
-                                }
-                            };
-
-                            let Protocol::Ping { client_timestamp, nonce, } = command else {
+pub async fn spawn(port: u16) -> crate::Result<()> {
+    let socket = net::DualStackLocalSocket::new(port)?;
+    let v4_addr = socket.local_ipv4_addr()?;
+    let v6_addr = socket.local_ip6_addr()?;
+    tokio::spawn(
+        async move {
+            // Initialize a buffer for the UDP packet. We use the maximum size of a UDP
+            // packet, which is the maximum value of 16 a bit integer.
+            let mut v4_buf = vec![0; 1 << 16];
+            let mut v6_buf = vec![0; 1 << 16];
+            let mut output_buf = Vec::new();
+
+            loop {
+                tracing::debug!(%v4_addr, %v6_addr, "awaiting qcmp packets");
+
+                match socket.recv_from(&mut v4_buf, &mut v6_buf).await {
+                    Ok((size, source)) => {
+                        let received_at = chrono::Utc::now().timestamp_nanos();
+                        let contents = match source {
```

Comment:
Since we get a `SocketAddr` from `recv_from`, that actually ends up being a good indicator of which buffer to look at.

Reply:
This felt like a good approach to me, as then we don't have to create a vec on each packet either.

Reply:
Yeah I like this approach a lot 👍

Reply:
Sweet!

Anything I can do to make this more easily mergable?

I'm happy to update `ep/agent-service` to `main` and resolve conflicts so that this is more easily merged, if you are swamped with other things.

Reply:
Yeah I'm swamped right now, feel free to just push your commit to the PR